crates/compiler/builtins/roc/Dict.roc

```diff
@@ -208,3 +210,413 @@ insertIfVacant = \dict, key, value ->
+# It is 64bit and little endian specific currently.
+LowLevelHasher := { originalSeed : U64, state : U64 } has [
+         Hasher {
+             addBytes,
+             addU8,
+             addU16,
+             addU32,
+             addU64,
+             addU128,
+             addI8,
+             addI16,
+             addI32,
+             addI64,
+             addI128,
+             complete,
+             reset,
+         },
+     ]
+
+createLowLevelHasher : { seed ?U64 } -> LowLevelHasher
+createLowLevelHasher = \{ seed ? 0x526F_6352_616E_643F } ->
+    @LowLevelHasher { originalSeed: seed, state: seed }
+
+combineState : LowLevelHasher, { a : U64, b : U64, seed : U64, length : U64 } -> LowLevelHasher
+combineState = \@LowLevelHasher { originalSeed, state }, { a, b, seed, length } ->
+    tmp = wymix (Num.bitwiseXor wyp1 a) (Num.bitwiseXor seed b)
+    hash = wymix (Num.bitwiseXor wyp1 length) tmp
+
+    @LowLevelHasher { originalSeed, state: wymix state hash }
+
+complete = \@LowLevelHasher { state } -> state
+
+reset = \@LowLevelHasher { originalSeed } ->
+    @LowLevelHasher { originalSeed, state: originalSeed }
+
+addI8 = \hasher, i8 ->
+    addU8 hasher (Num.toU8 i8)
+addI16 = \hasher, i16 ->
+    addU16 hasher (Num.toU16 i16)
+addI32 = \hasher, i32 ->
+    addU32 hasher (Num.toU32 i32)
+addI64 = \hasher, i64 ->
+    addU64 hasher (Num.toU64 i64)
+addI128 = \hasher, i128 ->
+    addU128 hasher (Num.toU128 i128)
+
+# These implementations hash each value individually with the seed and then mix
+# the resulting hash with the state. There are other options that may be faster
+# like using the output of the last hash as the seed to the current hash.
+# I am simply not sure the tradeoffs here. Theoretically this method is more sound.
+# Either way, the performance will be similar and we can change this later.
+addU8 = \@LowLevelHasher { originalSeed, state }, u8 ->
+    seed = Num.bitwiseXor originalSeed wyp0
+    p0 = Num.toU64 u8
+    a =
+        Num.shiftLeftBy p0 16
+        |> Num.bitwiseOr (Num.shiftLeftBy p0 8)
+        |> Num.bitwiseOr p0
+    b = 0
+
+    combineState (@LowLevelHasher { originalSeed, state }) { a, b, seed, length: 1 }
+
+addU16 = \@LowLevelHasher { originalSeed, state }, u16 ->
+    seed = Num.bitwiseXor originalSeed wyp0
+    p0 = Num.bitwiseAnd u16 0xFF |> Num.toU64
+    p1 = Num.shiftRightZfBy u16 8 |> Num.toU64
+    a =
+        Num.shiftLeftBy p0 16
+        |> Num.bitwiseOr (Num.shiftLeftBy p1 8)
+        |> Num.bitwiseOr p1
+    b = 0
+
+    combineState (@LowLevelHasher { originalSeed, state }) { a, b, seed, length: 2 }
+
+addU32 = \@LowLevelHasher { originalSeed, state }, u32 ->
+    seed = Num.bitwiseXor originalSeed wyp0
+    p0 = Num.toU64 u32
+    p1 = Num.toU64 u32
+    a = Num.shiftLeftBy p0 32 |> Num.bitwiseOr p1
+    b = Num.shiftLeftBy p1 32 |> Num.bitwiseOr p0
```

Comment:
Looks like `p0` and `p1` will be the same, and `a` and `b` will be the same - is that intentional? Asking because it seems like it might be a typo; if not, can the redundant variables be removed?

Reply:
Originally I wrote it to mirror the format of the other cases, but yeah in the u32 case, it is just repeated a bunch. updated.